drivers/serial/uart_mcux_lpuart.c

```diff
@@ -855,8 +855,7 @@ static void mcux_lpuart_isr(const struct device *dev)
 			data->tx_poll_stream_on = false;
 			mcux_lpuart_pm_policy_state_lock_put(dev);
 		}
-		assert(LPUART_ClearStatusFlags(config->base,
-			kLPUART_TransmissionCompleteFlag) == 0U);
+		LPUART_ClearStatusFlags(config->base, kLPUART_TransmissionCompleteFlag);
```

Comment:
what's the problem with leaving the assert on?

Reply:
Assertions should generally be turned off in production/release as they are a development tool. 

From the description of CONFIG_ASSERT (zephyr\subsys\debug\Kconfig):

> 	  Disabling this option will cause assertions to compile to nothing,
>	  improving performance and system footprint.

By default, CONFIG_ASSERT is only enabled if CONFIG_TEST is enabled.

No matter what, assertions should never be **necessary** for the correct functionality.

Reply:
I believe you either thread the error case (because that's why the assert was added, in the first place) or you explain (add a comment in code) why you think there can't be any issues here (see the implementation of [LPUART_ClearStatusFlags()](https://github.com/nxp-mcuxpresso/mcux-sdk/blob/main/drivers/lpuart/fsl_lpuart.c#L903)).

Reply:
Thank you for the link:

> * return 0 succeed, others failed.
> * retval kStatus_LPUART_FlagCannotClearManually The flag can't be cleared by this function but
> *         it is cleared automatically by hardware.
> * retval kStatus_Success Status in the mask are cleared.

To me this looks like a return value != 0 would not have been a problem under **any** circumstances and the assert seems to be overkill. A return value of `kStatus_LPUART_FlagCannotClearManually ` would have meant that this call was not necessary, while the assert would have actually been harmfull if `kStatus_Success` was not defined to be 0.

Besides, the Function LPUART_ClearStatusFlags() is called six times in this driver - in the other four cases, the return value is ignored as well. I don't think it is necessary to treat it differently here.

Reply:
Looking, again, in `LPUART_ClearStatusFlags()`, I see that here https://github.com/nxp-mcuxpresso/mcux-sdk/blob/main/drivers/lpuart/fsl_lpuart.c#LL909C45-L909C45 we take only the clearable flags (which are here: https://github.com/nxp-mcuxpresso/mcux-sdk/blob/main/drivers/lpuart/fsl_lpuart.h#L210) so, in this case, `kLPUART_TransmissionCompleteFlag` is not clearable, so the `mask = 0`, so this function does nothing.
Therefore, in IMO, the call to `LPUART_ClearStatusFlags` should be completely removed.

cc: @danieldegrasse 

>Besides, the Function LPUART_ClearStatusFlags() is called six times in this driver - in the other four cases, the return >value is ignored as well. I don't think it is necessary to treat it differently here.

The other calls should be check too.

Reply:
That's an interesting observation about `kLPUART_TransmissionCompleteFlag` - if I understand this correctly, the code would never have worked with `LPUART_ClearStatusFlags(config->base, kLPUART_TransmissionCompleteFlag)` returning `kStatus_LPUART_FlagCannotClearManually` which is != 0 and would have triggered the assertion. Apparently, nobody ever writes to an UART in an application with enabled power management?

Reply:
> That's an interesting observation about `kLPUART_TransmissionCompleteFlag` - if I understand this correctly, the code would never have worked with `LPUART_ClearStatusFlags(config->base, kLPUART_TransmissionCompleteFlag)` returning `kStatus_LPUART_FlagCannotClearManually` which is != 0 and would have triggered the assertion. 

No, the call `LPUART_ClearStatusFlags(config->base, kLPUART_TransmissionCompleteFlag)` always returns `kStatus_Success`, because [here](https://github.com/nxp-mcuxpresso/mcux-sdk/blob/main/drivers/lpuart/fsl_lpuart.c#L909) the mask get's 0 and [here](https://github.com/nxp-mcuxpresso/mcux-sdk/blob/main/drivers/lpuart/fsl_lpuart.c#L929) goes on else.

>Apparently, nobody ever writes to an UART in an application with enabled power management?

This, I really don't know. :(

Reply:
ah yet, you're right.

Reply:
Good point @iuliana-prodan- I was not aware that the TC flag could not be cleared by software- so it looks like we can remove the `LPUART_ClearStatusFlags(config->base, kLPUART_TransmissionCompleteFlag);`

As for the constant interrupt issue, I expect that the call `LPUART_ClearStatusFlags(config->base, kLPUART_IdleLineFlag);` being within an assert might be the culprit. @kt-peters, are you using the UART ASYNC API? If so, that call being optimized away would cause issues.

> Apparently, nobody ever writes to an UART in an application with enabled power management?

The implementation of the UART ASYNC API and support for power management with UART were both verified using Zephyr testcases, which would set `CONFIG_TEST=y` and `CONFIG_ASSERT=y`. So that likely explains why this issue hasn't be caught by testing.

Reply:
Based on @iuliana-prodan's observation, I have removed the attempt to clear the unclearable `kLPUART_TransmissionCompleteFlag` flag. 

@danieldegrasse: Yes, I'm using the `UART_ASYNC_API`, so my problem actually occurred when the `kLPUART_IdleLineFlag` was not cleared. I changed this line as well as I expected it to cause the same problems for the same reason (supposedly critical function call inside an `assert()`).

Reply:
@iuliana-prodan, are you satisfied with this solution?

Reply:
Yes, glad we found the exact root cause of your issue!